### PR TITLE
Fixes #30998 - Global Registration - Secure connection

### PR DESCRIPTION
**Pull request by stejskalleos**
Use Foreman `Setting[:ssl_ca_file]` certificate for CURL API calls.

**Preconditions:**
* Setup with Katello (https://github.com/Katello/katello/pull/8965), `activation_keys` & SSL
* In Settings: `Foreman URL` with https
* Generic OS with assigned `registration template`

**Commands:**
```
# For CentOS & RHEL
curl -X GET "http://foreman.example.com/register?organization_id=1&location_id=2&activation_key=sample-key" --user admin:changeme | bash

# Other OS:
curl -X GET "http://foreman.example.com/register?organization_id=1&location_id=2&operatingsystem_id=OS_ID" --user admin:changeme | bash
```

**Tested on:**
OS | Version
------------ | -------------
Fedora | 31
CentOS | 7
RHEL | 7.8
Debian | 10
Ubuntu | 18.04

**Notes:**
* Scenarios with proxy are not covered yet.

---

PR is part of **Simple & automatic host registration WF**, see:
* [RFC](https://community.theforeman.org/t/rfc-simple-automatic-host-registration-wf/19588)
* [Redmine (this PR)](https://projects.theforeman.org/issues/30441)
* [Redmine (Parent task)](https://projects.theforeman.org/issues/30440)

### CODE REVIEW

File: app/controllers/concerns/foreman/controller/registration.rb [hunk 1]
```diff
@@ -18,6 +18,7 @@ def global_registration_vars
     location = Location.authorized(:view_locations).find(params['location_id']) if params['location_id'].present?
     host_group = Hostgroup.authorized(:view_hostgroups).find(params['hostgroup_id']) if params["hostgroup_id"].present?
     operatingsystem = Operatingsystem.authorized(:view_operatingsystems).find(params['operatingsystem_id']) if params["operatingsystem_id"].present?
+    ssl_ca_file = File.read(Setting[:ssl_ca_file]) rescue nil
```
Comment (ares) [inline app/controllers/concerns/foreman/controller/registration.rb]: could we rename the variable to `ssl_ca_cert` please? this is the content of the file

File: app/controllers/concerns/foreman/controller/registration.rb [hunk 2]
```diff
@@ -27,7 +28,8 @@ def global_registration_vars
                    organization: organization,
                    location: location,
                    hostgroup: host_group,
-                   operatingsystem: operatingsystem })
+                   operatingsystem: operatingsystem,
+                   ssl_ca_file: ssl_ca_file })
```
Comment (ares) [inline app/controllers/concerns/foreman/controller/registration.rb]: `ssl_ca_cert` here too

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 3]
```diff
@@ -25,8 +25,13 @@ if [ -f /etc/os-release ] ; then
   . /etc/os-release
 fi
 
+SSL_CA_FILE=$(mktemp --suffix .crt)
+tee $SSL_CA_FILE << CA
+<%= @ssl_ca_file %>
```
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: `@ssl_ca_cert` please

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 4]
```diff
@@ -57,7 +62,7 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
     yum localinstall $CONSUMER_RPM -y
     rm -f $CONSUMER_RPM
 
-    subscription-manager register --org=<%= @organization.label %> --activationkey="sample-key"
+    subscription-manager register --org=<%= @organization.label %> --activationkey="<%= @activation_key %>"
```
Comment (ares) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: I know it can be frustrating, but could you please split this to a separate PR? This one is just about deploying the SSL cert for the registration purpose. This is a valid enhancement that will be easy to test and merge.

Comment (timogoebel) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: Can we detect if the katello plugin is actually available?

Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: Yes, we do that: https://github.com/theforeman/foreman/blob/a2c96913c1873dfe00630a6c097ac5fea4823e6f/app/views/unattended/provisioning_templates/registration/global_registration.erb#L48

File: app/controllers/concerns/foreman/controller/registration.rb (hunk 1 above)
Comment (ares) [inline app/controllers/concerns/foreman/controller/registration.rb]: also should we hard fail in the case the file can't be read? this effectively breaks the registration process, curl will fail for sure, so it's better to fail earlier and provide explanatory error message

**Review by ares — CHANGES_REQUESTED** [review]
Few initial comments, I'll proceed with testing since I don't expect big changes.

File: app/controllers/concerns/foreman/controller/registration.rb (hunk 1 above)
Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration.rb]: But what if somebody don't want to use https? There can be users like that.

Comment (timogoebel) [reply app/controllers/concerns/foreman/controller/registration.rb]: I believe we need some error handling here, maybe with a macro?

Comment (ares) [reply app/controllers/concerns/foreman/controller/registration.rb]: I think we should ignore the error only in the case if file can't be read and the current request is not https. If it is https, we should consider this as a hard failure. All the links we generate using path helpers will inherit the protocol too.

> I believe we need some error handling here, maybe with a macro?

do we need macros if the endpoint fails even before the template is rendered? or could you describe a bit better what you have in mind?

Comment (timogoebel) [reply app/controllers/concerns/foreman/controller/registration.rb]: I believe we should use the chance to "upgrade" the connection to https, I would not spend any time to make this intelligently work with both HTTP and HTTPs.

What I meant with error handling is: If the file can't be read, we need to give the user a clear error message that this failed and why it failed. And my idea was, that we could actually make this a macro (that would print the certificate). And only if the macro is used (it could come in handy during normal provisioning as well, the call home call for example) and there is an error reading the file show an error message. If a user does not need the certificate data, we don't need to make this a requirement to render the template.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 4 above)
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: https://github.com/theforeman/foreman/pull/8048

File: app/controllers/concerns/foreman/controller/registration.rb (hunk 1 above)
Comment (ares) [reply app/controllers/concerns/foreman/controller/registration.rb]: @stejskalleos convinced me so +1 to have something like `<%= foreman_ssl_certificate %>` that fails in case it can't be read. We can start enforcing HTTPS.

**Comment by stejskalleos** [comment]
Added new macro `<%= foreman_ssl_certificate %>`

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 5]
```diff
@@ -25,8 +25,13 @@ if [ -f /etc/os-release ] ; then
   . /etc/os-release
 fi
 
+SSL_CA_CERT=$(mktemp --suffix .crt)
+tee $SSL_CA_CERT << CA
+<%= foreman_ssl_certificate %>
```
Comment (timogoebel) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: Can we use the `save_to_file` macro?

Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: We could do something like this:
```ruby
<%
ca_cert = "/tmp/foreman_#{rand_hex(5)}.crt"
%>

<%= save_to_file(ca_cert, foreman_ssl_certificate) %>

curl --cacert <%= ca_cert %> --request POST
```
But that's the same as now, but just with more ruby.

Comment (timogoebel) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: I personally prefer the shell way, that's easier to read and comprehend. I was just wondering, we might have wanted to use our own macros here (dogfooding).

File: lib/foreman/renderer/scope/macros/base.rb [hunk 6]
```diff
@@ -334,6 +334,17 @@ def gem_version_compare(first, second)
             Gem::Version.new(first.to_s) <=> Gem::Version.new(second.to_s)
           end
 
+          apipie :method, 'Returns content of Foreman SSL certificate' do
+            example 'foreman_ssl_certificate # => Content of file'
+          end
+          def foreman_ssl_certificate
+            if File. exist?(Setting[:ssl_ca_file])
```
Comment (timogoebel) [inline lib/foreman/renderer/scope/macros/base.rb]:
```suggestion
            if File.exist?(Setting[:ssl_ca_file])
```

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 7]
```diff
@@ -25,8 +25,13 @@ if [ -f /etc/os-release ] ; then
   . /etc/os-release
 fi
 
+SSL_CA_CERT=$(mktemp --suffix .crt)
+tee $SSL_CA_CERT << CA
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: I wonder why you use tee instead of `cat`. Tee also outputs it to stdout but you don't really need that, right?

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 8]
```diff
@@ -25,8 +25,13 @@ if [ -f /etc/os-release ] ; then
   . /etc/os-release
 fi
 
+SSL_CA_CERT=$(mktemp --suffix .crt)
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: I think `--suffix` is a GNU extension and won't work on BSDs.

I think it's also good to clean up after yourself and do a `rm $SSL_CA_CERT` in the end.

File: lib/foreman/renderer/configuration.rb [hunk 9]
```diff
@@ -42,7 +42,8 @@ class Configuration
         :number_with_delimiter,
         :number_with_precision,
         :number_to_human_size,
-        :gem_version_compare
+        :gem_version_compare,
+        :foreman_ssl_certificate
```
Comment (ekohl) [inline lib/foreman/renderer/configuration.rb]: Let's use a trailing comma to having to modify it in the future when another line is added.
```suggestion
        :foreman_ssl_certificate,
```

File: lib/foreman/renderer/scope/macros/base.rb [hunk 10]
```diff
@@ -334,6 +334,17 @@ def gem_version_compare(first, second)
...
+              File.read(Setting[:ssl_ca_file])
+            else
+              raise Foreman::Exception.new("Foreman SSL certificate '#{Setting[:ssl_ca_file]}' not found")
```
Comment (ekohl) [inline lib/foreman/renderer/scope/macros/base.rb]: Are exceptions a good idea in macros? You can't catch them. Also note that you do not need to run Foreman using SSL (like in development setups).

`ssl_ca_file` is listed as a blank attr so it's a valid scenario to encounter. `File.exist?('')` returns `false` and you'll end up here in your development setup.

Returning an empty string is actually fine IMHO.

File: lib/foreman/renderer/scope/macros/base.rb [hunk 11]
```diff
@@ -334,6 +334,17 @@ def gem_version_compare(first, second)
             Gem::Version.new(first.to_s) <=> Gem::Version.new(second.to_s)
           end
 
+          apipie :method, 'Returns content of Foreman SSL certificate' do
```
Comment (ekohl) [inline lib/foreman/renderer/scope/macros/base.rb]: This is incorrect. The setting is defined as this:
https://github.com/theforeman/foreman/blob/ba3da753f5c68ea487dee6834fb39e362e2e9e1f/app/models/setting/auth.rb#L13

So the `ssl_ca_file` setting only contains the CA file for the Smart Proxies. In a default setup we happen to use the same CA, but if you're using a custom certificate on Foreman itself then you're not guaranteed to be able to connect.

File: lib/foreman/renderer/scope/macros/base.rb [hunk 12]
```diff
@@ -334,6 +334,17 @@ def gem_version_compare(first, second)
...
+            example 'foreman_ssl_certificate # => Content of file'
```
Comment (ekohl) [inline lib/foreman/renderer/scope/macros/base.rb]: Perhaps the example you're using in the template is actually helpful?
```bash
CA_FILE=$(mktemp)
cat > $CA_FILE <<CA_CONTENT
<%= foreman_ssl_certificate %>
CA_CONTENT
curl --cacert $CA_FILE https://smart-proxy.example.com:8443
```

File: lib/foreman/renderer/scope/macros/base.rb [hunk 13]
```diff
@@ -334,6 +334,17 @@ def gem_version_compare(first, second)
...
+          end
+          def foreman_ssl_certificate
```
Comment (ekohl) [inline lib/foreman/renderer/scope/macros/base.rb]: As pointed out elsewhere: the name of the macro doesn't match the content of the macro. Foreman doesn't have any setting now to figure out its own CA since that's only served by Apache or whatever webserver you're using.

Apache can be configured to send `SSL_SERVER_CERT` (the PEM encoded server certificate). It does have the downside that a certificate can be rotated while the CA typically isn't. So there are more failure scenarios to using it.

https://httpd.apache.org/docs/current/mod/mod_ssl.html

File: test/unit/foreman/renderer/renderers_shared_tests.rb [hunk 14]
```diff
@@ -205,6 +205,19 @@ module RenderersSharedTests
...
     end
 
+    test "foreman_ssl_certificate" do
+      cert_path = Rails.root.join('test/static_fixtures/certificates/example.com.crt')
+      Setting[:ssl_ca_file] = cert_path
+      source = OpenStruct.new(content: '<%= foreman_ssl_certificate %>')
+
+      assert_equal(renderer.render(source, @scope), File.read(cert_path))
+
+      Setting[:ssl_ca_file] = 'not-existing-file'
```
Comment (ekohl) [inline test/unit/foreman/renderer/renderers_shared_tests.rb:216]: IMHO it's a good habit to separate this out into its own test. Rough guideline is to have one assertion per test.

File: lib/foreman/renderer/scope/macros/base.rb (hunk 13 above)
Comment (ares) [reply lib/foreman/renderer/scope/macros/base.rb]: There's another downside. I couldn't workaround so far, on debian system, curl --cacert does not accept such certificate. It may be my hands too but it seems it needs full path (multiple certs). From your other comments, I'm not sure if you're OK with using the suggested setting and ask for changing the description or you request changing the whole thing and get the actual cert for HTTP headers. Could you please clarify?

File: lib/foreman/renderer/scope/macros/base.rb (hunk 10 above)
Comment (ares) [reply lib/foreman/renderer/scope/macros/base.rb]: I believe we correctly rescue from exceptions during rendering and I'm sure we use them https://github.com/theforeman/foreman/blob/develop/lib/foreman/renderer/errors.rb

There was already a discussion on this, we agreed to upgrade the communication to https at this stage. We may reconsider and start supporting HTTP in future, if we see production setups, where it's demanded. But dev setup does not qualify for keeping unsecure communication I think.

File: lib/foreman/renderer/scope/macros/base.rb (hunk 13 above)
Comment (ekohl) [reply lib/foreman/renderer/scope/macros/base.rb]: > curl --cacert does not accept such certificate

With such, do you mean the server certificate, not the CA?

> From your other comments, I'm not sure if you're OK with using the suggested setting and ask for changing the description or you request changing the whole thing and get the actual cert for HTTP headers. Could you please clarify?

I think using the current setting is a good start and will work for a lot of cases. but `foreman_ssl_certificate` is an imprecise name that can only create confusion in its current implementation. I'm struggling to come up with a better name/description.

File: lib/foreman/renderer/scope/macros/base.rb (hunk 10 above)
Comment (ekohl) [reply lib/foreman/renderer/scope/macros/base.rb]: I'd be OK with only supporting that, but then I think we should improve the error reporting to make it actionable for the user. `Foreman SSL certificate` is again a bad description. As a user I would not connect this to the actual setting since the descriptions do not match.

Perhaps it also makes sense to special case the blank value and raise an exception for that? `Foreman SSL certificate '' not found` may cause frustration while ```No SSL CA file is configured. Set the `ssl_ca_file` setting``` (or similar) is more instructive in how to solve it as a user.

File: lib/foreman/renderer/configuration.rb (hunk 9 above)
Comment (ekohl) [reply lib/foreman/renderer/configuration.rb]: Turns out Rubocop dislikes this. We should solve that elsewhere.

File: lib/foreman/renderer/scope/macros/base.rb (hunk 13 above)
Comment (ares) [reply lib/foreman/renderer/scope/macros/base.rb]: > With such, do you mean the server certificate, not the CA?

Yes

> I think using the current setting is a good start and will work for a lot of cases. but foreman_ssl_certificate is an imprecise name that can only create confusion in its current implementation. I'm struggling to come up with a better name/description.

Ok. If we know, this should in the longer term be the foreman cert or foreman ca cert, I'd keep the name or call it `foreman_server_ca_cert`. User may want to customize this, for now they can edit the template and put it there. I guess we won't improve it until we refactor the whole cert handling.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 8 above)
Comment (ares) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: Are all our templates BSD compatible? I know some systems are picky about the suffix, but `curl` probably does not care. If that's the case, let's drop that.

File: lib/foreman/renderer/scope/macros/base.rb (hunk 13 above)
Comment (ekohl) [reply lib/foreman/renderer/scope/macros/base.rb]: Perhaps open an issue and add a TODO in the code with that issue? One mitigating factor is that custom certs are often used to get a trusted certificate (like Let's Encrypt) or some corporate CA so it's probably not too bad. There's a decent chance those certificates are already in place on the target machine.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 8 above)
Comment (ekohl) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: I think we did our best and I tried to spot all things I could. If curl doesn't care, there's not really a benefit to using it now.

File: lib/foreman/renderer/scope/macros/base.rb (hunk 11 above)
Comment (timogoebel) [reply lib/foreman/renderer/scope/macros/base.rb]: Why don‘t we simply add a dedicated setting for the ca cert?

File: lib/foreman/renderer/scope/macros/base.rb (hunk 13 above)
Comment (timogoebel) [reply lib/foreman/renderer/scope/macros/base.rb]: I think we simply need the ca cert here. I think it‘s reasonable to introduce a setting for this, we can iterate on that later. The macro should clearly state that it returns the server ca cert.
The current approach is just technical debt that will do more harm than good. Let‘s not get carried away here.

**Review by ares — CHANGES_REQUESTED** [review]
I have tested this with both Katello and non-katello workflow, on debian and centos, works as expected :+1: needs only rebase (conflict and to run katello tests). I'm creating a follow up issue for the setting. Imho good to merge after the rebase.

**Review by ares — APPROVED** [review]
Thanks @stejskalleos, merging!

File: lib/foreman/renderer/scope/macros/base.rb [hunk 15]
```diff
@@ -334,6 +334,22 @@ def gem_version_compare(first, second)
...
           end
 
+          apipie :method, "Returns content of 'SSL CA file' configured in Settings > Authentication" do
+            example "SSL_CA_CERT=$(mktemp)
+                     cat > $SSL_CA_CERT <<CA_CONTENT
+                     <%= foreman_server_ca_cert %>
+                     CA_CONTENT
+                     curl --cacert $SSL_CA_CERT https://smart-proxy.example.com:8443"
+          end
+          def foreman_server_ca_cert
```
Comment (ekohl) [inline lib/foreman/renderer/scope/macros/base.rb:344]: I still think this doesn't really match the description.

Comment (ares) [reply lib/foreman/renderer/scope/macros/base.rb:344]: FTR being addressed at https://github.com/theforeman/foreman/pull/8074
